swish-test updates and tilde-expansion for filesystem operations

owaddell opened the pull request:
**Proposed changes**

Add `swish-test` options to make it easier for folks to generate a text-based summary of prior test results or to re-run prior tests that failed or were skipped. Handle limited (current-user) form of tilde expansion for filesystem operations.

* `swish-test`: allow `--progress` in combination with `--harvest` or `--load-results`.
* `swish-test`: add `--rerun` option
* `swish-test`: include exit reason when reporting failure to load profile data
* add `osi_get_home_directory`
* support `~` and paths with `~/` prefix in filesystem procedures

owaddell-beckman committed:
swish-test: allow --progress with --harvest or --load-results

Allow --progress with --harvest or --load-results so folks can
obtain a concise text summary of results from earlier test runs.

Fix a bug where long suite names could bleed into the suite status
in --progress suite.

owaddell-beckman committed:
swish-test: explain why profile data fails to load

If we attempt to profile code containing (include "~/tmp/debug.ss"),
Chez Scheme's include expands the tilde and finds the file. Swish's
$profile-load doesn't (yet) handle the tilde and raises an exception,
but now we at least report the reason.

owaddell-beckman committed:
add osi_get_home_directory

owaddell-beckman committed:
add limited tilde expansion for filesystem paths

Expand "~" and paths beginning with "~/" by using osi_get_home_directory.
While this handles the most common cases, it does not handle paths that
start with "~user/" as Chez Scheme's internal S_malloc_pathname does.

owaddell-beckman committed:
bolster argument checks in filesystem procedures

owaddell-beckman committed:
fix mat load-results to preserve result order

owaddell-beckman committed:
swish-test: preserve --harvest suite order

Preserve command-line suite ordering when processing --harvest option.
This becomes salient when we allow --progress with --harvest.

owaddell-beckman committed:
swish-test: add --rerun option

burgerrg commented on src/swish/swish-test.ms:
```diff
@@ -228,8 +240,9 @@
+  (define subdir (uuid->string (osi_make_uuid)))
   (define test-file
-    (write-test-file "test-specified.ms"
+    (write-test-file (path-combine "test-specified.ms")
```
Are you missing the path argument to path-combine?

owaddell replied:
Well spotted. I meant to use the subdir to check the case where we `--rerun` all results in a particular directory.

Pushed a fix for that along with some proposed breaking changes.

owaddell-beckman committed:
breaking change: no automatic .ms suite extension

swish-test locates tests by searching specified directories or by loading
specified files, but no longer adds an automatic .ms suffix when checking
whether a specified file exists.

Automatically adding a .ms extension led to confusion in cases like the
following where the user intends to run the foo test defined in tests.ms,
but instead runs all tests in tests.ms and the foo.ms suite that happens
to be in the same directory:

  swish-test tests.ms foo

The user could have added a -t flag to ensure that swish-test treats foo
as a test name:

  swish-test tests.ms -t foo

Eliminating the automatic .ms suffix makes it clearer where the -t flag
might be needed.

owaddell-beckman committed:
breaking: change --harvest and --load-result exit status

Make --harvest and --load-result exit with non-zero status when processing
results that are incomplete or that contain failed tests, skipped tests,
or no passing tests. The aim is to generate a non-zero exit status when the
original test run that produced those results would have done so.

payneca committed:
test: prevent gc from closing port in http mats

burgerrg reviewed (approved):
Thanks for these useful changes!

laqrix reviewed (approved):
Looks good. Thanks Oscar!

owaddell commented:
Thanks for the reviews, folks!
